src/mono/msbuild/AndroidAppBuilder/Templates/runtime-android.c

```diff
@@ -0,0 +1,197 @@
+    strlcpy (filename, name, sizeof (filename));
+    if (!has_extension) {
+        strlcat (filename, ".dll", sizeof (filename));
+    }
+
+    if (culture && strcmp (culture, ""))
+        res = snprintf (path, sizeof (path) - 1, "%s/%s/%s", bundle_path, culture, filename);
+    else
+        res = snprintf (path, sizeof (path) - 1, "%s/%s", bundle_path, filename);
+    assert (res > 0);
+
+    struct stat buffer;
+    if (stat (path, &buffer) == 0) {
+        MonoAssembly *assembly = mono_assembly_open (path, NULL);
+        assert (assembly);
+        return assembly;
+    }
+    return NULL;
+}
+
+static MonoAssembly*
+assembly_preload_hook (MonoAssemblyName *aname, char **assemblies_path, void* user_data)
+{
+    const char *name = mono_assembly_name_get_name (aname);
+    const char *culture = mono_assembly_name_get_culture (aname);
+    return load_assembly (name, culture);
+}
+
+char *
+strdup_printf (const char *msg, ...)
+{
+    va_list args;
+    char *formatted = NULL;
+    va_start (args, msg);
+    vasprintf (&formatted, msg, args);
+    va_end (args);
+    return formatted;
+}
+
+static MonoObject *
+fetch_exception_property (MonoObject *obj, const char *name, bool is_virtual)
+{
+    MonoMethod *get = NULL;
+    MonoMethod *get_virt = NULL;
+    MonoObject *exc = NULL;
+
+    get = mono_class_get_method_from_name (mono_get_exception_class (), name, 0);
+    if (get) {
+        if (is_virtual) {
+            get_virt = mono_object_get_virtual_method (obj, get);
+            if (get_virt)
+                get = get_virt;
+        }
+
+        return (MonoObject *) mono_runtime_invoke (get, obj, NULL, &exc);
+    } else {
+        printf ("Could not find the property System.Exception.%s", name);
+    }
+
+    return NULL;
+}
+
+static char *
+fetch_exception_property_string (MonoObject *obj, const char *name, bool is_virtual)
+{
+    MonoString *str = (MonoString *) fetch_exception_property (obj, name, is_virtual);
+    return str ? mono_string_to_utf8 (str) : NULL;
+}
+
+void
+unhandled_exception_handler (MonoObject *exc, void *user_data)
+{
+    MonoClass *type = mono_object_get_class (exc);
+    char *type_name = strdup_printf ("%s.%s", mono_class_get_namespace (type), mono_class_get_name (type));
+    char *trace = fetch_exception_property_string (exc, "get_StackTrace", true);
+    char *message = fetch_exception_property_string (exc, "get_Message", true);
+    
+    LOG_ERROR("UnhandledException: %s %s %s", type_name, message, trace);
+
+    free (trace);
+    free (message);
+    free (type_name);
+    exit (1);
+}
+
+void
+log_callback (const char *log_domain, const char *log_level, const char *message, mono_bool fatal, void *user_data)
+{
+    LOG_INFO ("(%s %s) %s", log_domain, log_level, message);
+    if (fatal) {
+        LOG_ERROR ("Exit code: %d.", 1);
+        exit (1);
+    }
+}
+
+int
+mono_mobile_runtime_init (void)
+{
+    // for now, only Invariant Mode is supported (FIXME: integrate ICU)
```

Comment:
We will have platform level support for ICU on android.  You should be able to remove this.

Reply:
well, I guess we can remove it once it's implemented then 🙂

Reply:
That wasn't the right statement... Android provides platform level support for ICU.  It should already work.

Reply:
yeah, looks like it works